### PR TITLE
Flashflow

### PR DESCRIPTION
**Pull request by pastly**

### CODE REVIEW

File: proposals/ideas/xxx-flashflow.txt [hunk 1]
```diff
@@ -295,63 +291,40 @@ section:
...
-(successful or otherwise), the coordinator records the results and repeats the
-process. If all relays have been measured (successfully or otherwise) in the
-last 24 hours, the coordinator waits for a new consensus. If the new consensus
-lacks relays that need measuring, it continues waiting.
+The short term implementation of measurement scheduling will be simplier than
+the long term one due to (1) there only being one FlashFlow deployment, and
+(2) there being very few relays that support being measured by FlashFlow. In
+fact the FF coordinator will maintain a list of the relays that actually
+support being measured, have opted in to being measured, and will only measure
+them.
```
Comment (robgjansen) [inline proposals/ideas/xxx-flashflow.txt]: I think you mean this:

```
In fact, the FF coordinator will maintain a list of the relays that actually
support and have opted in to being measured, and we will only measure them.
```

File: proposals/ideas/xxx-flashflow.txt [hunk 2]
```diff
@@ -295,63 +291,40 @@ section:
...
+
+The coordinator divides time into a series of 24 hour periods, commonly
+referred to as days. Each period has measurement slots that are longer than a
+measurement lasts (30s), say 60s, to account for handshaking and recovery
+after a measurement is completed. Thus with 60s slots there's 1,440 slots in a
+day.
+
+At the start of each day the coordinator consider the list of relays that have
```
Comment (robgjansen) [inline proposals/ideas/xxx-flashflow.txt]: considers

File: proposals/ideas/xxx-flashflow.txt [hunk 3]
```diff
@@ -295,63 +291,40 @@ section:
...
+opted in to being measured.  From this list of relays, it repeatedly takes the
+relay with the largest existing capacity estimate. It selects a random slot.
+If the slot has existing relays assigned to it, the coordinator makes sure
+there is enough additional measurer capacity to handle this relay. If so, it
+assigns this relay to this slot. If not, it keeps picking new random slots
+until one has sufficient additional measurer capacity.
```
Comment (robgjansen) [inline proposals/ideas/xxx-flashflow.txt:326]: Collisions waste CPU resources, but I assume the likelihood of collision is small here. Maybe state the worst-case likelihood as a rough percentage to stop people from raising the issue? Or maybe that's non-trivial to compute, in which case nevermind...

**Review by robgjansen — COMMENTED** [review]
A couple of minor edits.

**Comment by pastly** [comment]
Todo list, will be updated

- [x] easy 3 things that @robgjansen said above
- [x] change author order if agreement reached
- [x] update torrc/consensus params: relay can get measured twice in any 24hr window, but that's it. 1 param: length of period. the 2-times is implicit
- [x] explicitly enumerate risks with the lack of authentication: anyone can measure any relay for X seconds Y times a day unless relay turns measurement off manually in their torrc
- [x] long term plan for co-located relays: detection or torrc option
- [x] acknowledgement of the speedtest that Rob did and the trac ticket

.

- [x] standardize usage of "TLS certificate"-like phrases
- [x] authentication security section
- [x] ~~specify that DoS conn defense disabled upon MSM_PARAMS_OK and until end of msm~~
- [x] specify better the measurer IPs are exempt from DoS defense
- [ ] once over
- [ ] remove instances of two spaces between period and next sentence

File: proposals/ideas/xxx-flashflow.txt [hunk 4]
```diff
@@ -0,0 +1,666 @@
...
+     target relay and verify that the cells echoed back match the ones sent.
+     During this time the relay caps the amount of background traffic it
+     transfers. Background and measurement traffic are handled separately at
+     the relay.
+  X. For every second during the measurement, the measurement nodes report to
+     the authority node how much traffic was echoed back.
+     The target relay also reports the amount of per-second background
+     (non-measurement) traffic.
+  X. The authority node sums the per-second reported throughputs into 30 sums
+     (one for each second) and calculates the median. This is estimated
```
Comment (robgjansen) [inline proposals/ideas/xxx-flashflow.txt]: "the estimated"

File: proposals/ideas/xxx-flashflow.txt [hunk 5]
```diff
@@ -0,0 +1,666 @@
...
+     capacity of the relay.
+
+FlashFlow performs a measurement of every relay according to a schedule
+described later in this document. Periodically it produces relay capacity
+estimates in the form of a v3bw file, which is suitable for consumption by a Tor
+directory and allows FlashFlow to be a voting "bandwidth authority."
+
+FlashFlow is more performant than Torflow: FlashFlow takes 5 hours to measure
+the entire existing Tor network from scratch (with 3 Gbit/s measurer capacity) while Torflow takes 2 days;
```
Comment (robgjansen) [inline proposals/ideas/xxx-flashflow.txt]: line extends too far into the right margin?

File: proposals/ideas/xxx-flashflow.txt [hunk 6]
```diff
@@ -0,0 +1,666 @@
...
+
+[XXX A paragraph summarizing the layout of this proposal]
+
+X. Deployment Stages
+
+FlashFlow's deployment shall be broken up into three stages.
+
+In the short term we want to implement a working FlashFlow measurement system.
+This requires code changes in little-t tor and an external FlashFlow codebase.
+The majority of the implementation work will be done in the short term, and the product is
```
Comment (robgjansen) [inline proposals/ideas/xxx-flashflow.txt]: line extends too far into the right margin?

**Review by robgjansen — COMMENTED** [review]
More edits.

File: proposals/ideas/xxx-flashflow.txt (hunk 3 above)
Comment (pastly) [reply proposals/ideas/xxx-flashflow.txt:326]: I put the full analysis in the proposal, but assuming:

- 200 relays opt in to being measured (a ton!)
- 1,440 slots per day (60 second slots for 30s measurement + slack time)
- only one relay can be in a slot (not true, but makes analysis easier)

then the number of random slot selections needed to pick a unique slot for each relay:

- median: 215
- 75th percentile: 218
- 95th percentile: 222
- maximum: 232

The number of CPU cycles wasted per day by the FF coord is inconsequential.

Comment (pastly) [reply proposals/ideas/xxx-flashflow.txt:326]: Removed this. Git history will hold on to it for now

**Pull request closed without merging**